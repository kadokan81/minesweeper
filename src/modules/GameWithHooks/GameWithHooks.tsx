import { FC, useState } from 'react';
import { GameArea } from '../../components/Game/GameArea';
import { GameOver } from '../../components/Game/GameOver';
import { Wrapper } from '../../components/Game/Wrapper';
import { Grid } from '../../components/Grid/Grid';
import { Scoreboard } from '../../components/Scoreboard';
import { Top } from '../../components/Top';
import { openCell } from '../../helpers/CellManipulation';
import {
	CellState,
	Coords,
	emptyFieldGenerator,
	Field,
	fieldGenerator,
} from '../../helpers/Field';
import { GameLevel, GameSettings, LevelNames } from './GameSettings';

const StaticField: Field = [
	[9, 2, 9, 1, 0, 0, 1, 1, 1, 1, 1, 1],
	[1, 2, 2, 2, 1, 0, 1, 9, 1, 1, 9, 1],
	[0, 0, 1, 9, 10, 0, 2, 2, 2, 1, 1, 1],
	[0, 0, 10, 10, 1, 0, 1, 9, 1, 2, 2, 2],
	[0, 1, 1, 2, 2, 9, 1, 1, 1, 0, 0, 0],
	[0, 1, 9, 3, 9, 2, 10, 0, 0, 2, 1, 1],
	[0, 2, 2, 4, 9, 2, 10, 1, 1, 1, 9, 1],
	[0, 1, 9, 2, 1, 1, 1, 9, 1, 2, 2, 2],
	[0, 1, 10, 10, 0, 0, 1, 1, 1, 1, 9, 1],
	[0, 1, 10, 10, 0, 0, 1, 1, 1, 1, 9, 1],
	[0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 9, 1],
	[0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 9, 1],
];

export const GameWithHooks: FC = () => {
	const [level, setLevel] = useState<LevelNames>('beginner');

	const [size, bombs] = GameSettings[level];

	const [playerField, setPlayerField] = useState<Field>(
		emptyFieldGenerator(size, CellState.hidden)
	);

	const gameField = fieldGenerator(size, bombs / (size * size));

	const onClick = (coords: Coords) => {
		const newPlayerField = openCell(coords, playerField, gameField);
		setPlayerField([...newPlayerField]);
	};
	const onChangeLevel = ({
		target: { value: level },
	}: React.ChangeEvent<HTMLSelectElement>) => {
		setLevel(level as LevelNames);
		const [size] = GameSettings[level as LevelNames];
		const newPlayerField = emptyFieldGenerator(size, CellState.hidden);
		setPlayerField([...newPlayerField]);
	};

	return (
		<Wrapper>
			<Top />
			<GameArea>
				<Scoreboard
					time='1200'
					mines={bombs + ''}
					levels={GameLevel as unknown as string[]}
					defaultLevel={level}
					onChangeLevel={onChangeLevel}
					onReset={() => {}}
				/>
				{/* <GameOver onClick={() => {}} isWin={true}></GameOver> */}
				<Grid onClick={onClick} onContextMenu={() => {}}>
					{playerField}
				</Grid>
			</GameArea>
		</Wrapper>
	);
};
