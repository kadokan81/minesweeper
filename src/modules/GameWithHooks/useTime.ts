import { useEffect, useState } from 'react';

export const useTime = (
	isGameStarted: boolean,
	gameIsOver: boolean
): [time: number, onReset: () => void] => {
	const [time, setTime] = useState(0);

	useEffect(() => {
		let interval: NodeJS.Timeout;
		if (isGameStarted) {
			interval = setTimeout(() => {
				setTime((time) => time + 1);
			}, 1000);

			if (gameIsOver) {
				clearInterval(interval);
			}
		}

		return () => {
			clearInterval(interval);
		};
	}, [isGameStarted, gameIsOver, time]);

	const onReset = () => setTime(0);

	return [time, onReset];
};
