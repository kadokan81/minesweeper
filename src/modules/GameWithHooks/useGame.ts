import {
	CellState,
	emptyFieldGenerator,
	fieldGenerator,
	Coords,
} from './../../helpers/Field';
import { useCallback, useEffect, useMemo, useState } from 'react';
import { Field } from '../../helpers/Field';
import { GameSettings, LevelNames } from './GameSettings';
import { openCell } from '../../helpers/openCell';
import { setFlag } from '../../helpers/setFlag';
import { log } from 'console';
import { useTime } from './useTime';

type returnType = {
	level: 'beginner' | 'intermediate' | 'expert';
	gameIsOver: boolean;
	isWin: boolean;
	settings: number[];
	playerField: Field;
	gameField: Field;
	flagCounter: number;
	time: number;
	onClick: (coords: Coords) => void;
	onChangeLevel: (level: LevelNames) => void;
	onReset: () => void;
	onContextMenu: (coords: Coords) => void;
};

export const useGame = (): returnType => {
	const [level, setLevel] = useState<LevelNames>('beginner');
	const [gameIsOver, setGAmeIsOver] = useState(false);
	const [isWin, setIsWin] = useState(false);
	const [isGameStarted, setIsGameStarted] = useState(false);

	const [size, bombs] = GameSettings[level];

	const [playerField, setPlayerField] = useState<Field>(
		emptyFieldGenerator(size, CellState.hidden)
	);
	const [gameField, setGameField] = useState<Field>(
		fieldGenerator(size, bombs / (size * size))
	);

	const [time, onResetTime] = useTime(isGameStarted, gameIsOver);

	useMemo(() => console.log(gameField), [gameField]);

	const [flagCounter, setFlagCounter] = useState(0);

	const onContextMenu = useCallback(
		(coords: Coords) => {
			!isGameStarted && setIsGameStarted(true);

			const [newPlayerField, isSolved, newFlagCounter] = setFlag(
				coords,
				playerField,
				gameField,
				flagCounter,
				bombs
			);
			setFlagCounter(newFlagCounter);
			if (isSolved) {
				setIsWin(isSolved);
				setGAmeIsOver(isSolved);
			}
			setPlayerField([...newPlayerField]);
		},
		[isGameStarted, gameIsOver, isWin, level, flagCounter]
	);

	const onReset = useCallback(() => {
		const newGameField = fieldGenerator(size, bombs / (size * size));
		const newPlayerField = emptyFieldGenerator(size, CellState.hidden);

		setGameField([...newGameField]);
		setPlayerField([...newPlayerField]);
		setGAmeIsOver(false);
		onResetTime();
		setIsGameStarted(false);
	}, [size, bombs]);

	const onClick = useCallback(
		(coords: Coords) => {
			!isGameStarted && setIsGameStarted(true);

			try {
				const [newPlayerField, isSolved, flagCounter] = openCell(
					coords,
					playerField,
					gameField
				);

				if (isSolved) {
					setIsWin(isSolved);
					setGAmeIsOver(isSolved);
				}
				setPlayerField([...newPlayerField]);
			} catch {
				setPlayerField([...gameField]);
				setGAmeIsOver(true);
				setIsWin(false);
			}
		},
		[isGameStarted, gameIsOver, isWin, level, flagCounter]
	);
	const onChangeLevel = (level: LevelNames = 'beginner') => {
		setLevel(level);

		const [size, bombs] = GameSettings[level];
		const newGameField = fieldGenerator(size, bombs / (size * size));
		const newPlayerField = emptyFieldGenerator(size, CellState.hidden);

		setGameField([...newGameField]);
		setPlayerField([...newPlayerField]);
		setGAmeIsOver(false);
		onResetTime();
		setIsGameStarted(false);
	};
	return {
		level,
		gameIsOver,
		isWin,
		settings: [size, bombs],
		playerField,
		gameField,
		flagCounter,
		onClick,
		onChangeLevel,
		onReset,
		onContextMenu,
		time,
	};
};
