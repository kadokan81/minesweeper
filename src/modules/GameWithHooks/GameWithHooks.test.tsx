import { screen } from '@storybook/testing-library';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { GameWithHooks } from './GameWithHooks';

describe('Game with hooks test cases', () => {
	describe('Render behavior', () => {
		it('render game field by default', () => {
			const { asFragment } = render(<GameWithHooks />);

			expect(screen.getAllByRole('cell')).toHaveLength(81);

			expect(asFragment).toMatchSnapshot();
		});
		it('onChange game level handler', () => {
			render(<GameWithHooks />);
			expect(screen.getAllByRole('cell')).toHaveLength(81);

			userEvent.selectOptions(screen.getByRole('combobox'), 'intermediate');
			expect(screen.getAllByRole('cell')).toHaveLength(256);

			userEvent.selectOptions(screen.getByRole('combobox'), 'expert');
			expect(screen.getAllByRole('cell')).toHaveLength(484);
		});
	});
});
