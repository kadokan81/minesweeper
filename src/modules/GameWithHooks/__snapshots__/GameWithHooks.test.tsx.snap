// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game with hooks test cases Render behavior render game field by default 1`] = `[Function]`;
